### PR TITLE
 [BEAM-7726] Implement State Backed Iterables in Go SDK

### PR DESCRIPTION
**Pull request by lostluck**
This PR does a few things towards adding State Backed Iterables
* Adds unit tests for the existing iterable protocols.
* Refactors the two disparate read loops in datasource.go into a single loop, and introduces helper methods to clean up repeated code.
  * This further helps prepare for [BEAM-490] where CoGBK becomes first class.
*   Renames code that is fairly State API centric to to State from SideInput. 
  * Most of the code was to help use the StateAPI for Side inputs, but it was really rather general.
  * Note: This only helps with Reading, and in particular Runner iterables from the state API, not doing anything towards Writing.
* Actually implement the State backed protocol (with unit test)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @robertwb, @lukecwik 
Note: This code diverges from the stated protocol for Iterables in beam_runner_api.proto [1] where multi-chunk iterables are prefixed with a 0. As implemented Go SDK, and Python SDKs [2] currently require a -1 instead. Either the SDKs need to change or the proto documentation should change to match.

[1] https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/beam_runner_api.proto#L582
[2] https://github.com/apache/beam/blob/master/sdks/python/apache_beam/coders/coder_impl.py#L702

**Comment by lostluck** [comment]
Run Go PreCommit

**Comment by lostluck** [comment]
cc: @herohde In case you've got thoughts on the refactoring.

Otherwise, I'm investigating the PreCommit failure. It looks like it's a Go Gradle thing WRT vendoring. It might be time to do away with it.

**Comment by lostluck** [comment]
cc: @ananvay

**Comment by lostluck** [comment]
R: @youngoli

**Comment by lostluck** [comment]
Run Go PreCommit

**Comment by robertwb** [comment]
> R: @robertwb, @lukecwik
> Note: This code diverges from the stated protocol for Iterables in beam_runner_api.proto [1] where multi-chunk iterables are prefixed with a 0. As implemented Go SDK, and Python SDKs [2] currently require a -1 instead. Either the SDKs need to change or the proto documentation should change to match.
> 
> [1] https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/beam_runner_api.proto#L582
> [2] https://github.com/apache/beam/blob/master/sdks/python/apache_beam/coders/coder_impl.py#L702

Ah, thanks for catching that. The docs are wrong (java uses -1 as well [3], and otherwise there'd be ambiguity in encoding an empty list). https://github.com/apache/beam/pull/9099

[3] https://github.com/apache/beam/blob/release-2.13.0/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/IterableLikeCoder.java#L101

File: sdks/go/pkg/beam/core/runtime/exec/datasource.go [hunk 1]
```diff
@@ -72,117 +79,129 @@ func (n *DataSource) Process(ctx context.Context) error {
...
-
-					for i := uint64(0); i < chunk; i++ {
-						value, err := cv.Decode(r)
-						if err != nil {
-							return errors.Wrap(err, "stream value decode failed")
-						}
-						buf = append(buf, *value)
-					}
+			case chunk == -1: // State backed iterable!
+				chunk, err := coder.DecodeVarInt(r)
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/datasource.go]: That's right this is the token's length, per the protocol in

File: sdks/go/pkg/beam/core/runtime/exec/datasource.go [hunk 2]
```diff
@@ -262,3 +282,60 @@ func (n *DataSource) Split(splits []int64, frac float32) (int64, error) {
...
+	if err == io.EOF {
+		if err := s.first.Close(); err != nil {
+			s.nextStream = nil
+			return nil, err
+		}
+		if s.nextStream == nil {
+			s.first = nil
+			return nil, io.EOF
+		}
+		s.first, err = s.nextStream.Open()
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/datasource.go]: That's right. For this use of concatStream, that's exactly what's happening. concatReStream and concatStream are general though, so any two re-streams can be concatenated together (including nested concat streams)

The continuations from statemgr.go took me a moment too, but fortunately that code is already nicely written.

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
It seems that between Go Gradle, and the version of go we have it run (go 1.10) it looks like the Beam Go SDK is no longer compatible with go 1.10.
In a separate client, I validated that the gradle build is much happier with go 1.12 (which gogradle manages by itself anyway). I'll resolve that first in a separate PR and then ensure this passes here before merging it in.

**Comment by lostluck** [comment]
Run Go PreCommit

**Comment by lostluck** [comment]
Run Go PostCommit

**Pull request merged**